(REGRESSION (265731@main): fast/scrolling/rtl-scrollbars-animation-property.html is a constant failure)

nmoucht opened the pull request:
#### 9320b50206384e7b857c586879952789b6d16aaf
<pre>
(REGRESSION (265731@main): fast/scrolling/rtl-scrollbars-animation-property.html is a constant failure)
<a href="https://bugs.webkit.org/show_bug.cgi?id=262930">https://bugs.webkit.org/show_bug.cgi?id=262930</a>
rdar://114349934

Reviewed by Cameron McCormack.

After <a href="https://commits.webkit.org/265731@main">https://commits.webkit.org/265731@main</a> ScrollbarThemeMac::isLayoutDirectionRTL did not have an
NSScrollerImp to query about being in rtl mode. AppKit only uses userInterfaceLayoutDirection for
rectForPart/expandedRectForPart, which are not used with UI-side compositing, so for now if we detect
that the scrollbar should not be registered (due to UI-side compositing being on), just query the scrollable
area to see if the scrollbar is in rtl mode.

* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::isLayoutDirectionRTL):

Canonical link: <a href="https://commits.webkit.org/269390@main">https://commits.webkit.org/269390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0d6a9d324ff827c9ee44844c1bc40ec1c9d6345

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23517 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20046 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21895 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22184 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21212 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18751 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24371 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18712 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25900 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19741 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19812 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23758 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20298 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17291 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19626 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23852 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2812 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20217 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
```diff
@@ -184,8 +184,11 @@ + (void)registerAsObserver
-    if (!scrollerImp)
+    if (!scrollerImp) {
+        if (!scrollbar.shouldRegisterScrollbar())
+            return scrollbar.scrollableArea().shouldPlaceVerticalScrollbarOnLeft() ? NSUserInterfaceLayoutDirectionRightToLeft : NSUserInterfaceLayoutDirectionLeftToRight;
```
When an `NSScrollerImp` is first created, we set its `userInterfaceLayoutDirection` based on the scrollable area, just like we are doing here. But can we return this regardless of whether an `NSScrollerImp` was created, and whether we should be registering it? I.e. replace the whole `#if PLATFORM(MAC)` section with this return statement that looks at the scrollable area? Or do we need to return false in these cases.

nmoucht replied:
I think it is more preferable to check the value of the scroller imp if it exists since we want to make sure it is in sync with the scrollable area value. Returning false for the case where we have no scroller imp (for the non ui-side compositing case) was how the previous code handled it so I think its fine to leave it like that.

heycam replied:
> since we want to make sure it is in sync with the scrollable area value

Does that mean there are times we know they may not be in sync? Or is that to defend against a bug we might not know about. I'll r+ but it still feels like there's an opportunity to simplify here.

nmoucht replied:
Yeah this would be to defend the case where the scrollable area value is right but the scroller imp's value is incorrect.

nmoucht committed:
(REGRESSION (265731@main): fast/scrolling/rtl-scrollbars-animation-property.html is a constant failure)
https://bugs.webkit.org/show_bug.cgi?id=262930
rdar://114349934

Reviewed by Cameron McCormack.

After https://commits.webkit.org/265731@main ScrollbarThemeMac::isLayoutDirectionRTL did not have an
NSScrollerImp to query about being in rtl mode. AppKit only uses userInterfaceLayoutDirection for
rectForPart/expandedRectForPart, which are not used with UI-side compositing, so for now if we detect
that the scrollbar should not be registered (due to UI-side compositing being on), just query the scrollable
area to see if the scrollbar is in rtl mode.

* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::isLayoutDirectionRTL):

Canonical link: https://commits.webkit.org/269390@main
